lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -568,49 +551,103 @@ def write(chunk)
         log_name = "projects/#{@project_id}/logs/#{log_name(
           tag, group_level_resource)}"
 
-        # Does the actual write to the cloud logging api.
-        if @use_grpc
-          labels_utf8_pairs = group_level_common_labels.map do |k, v|
-            [k.encode('utf-8'), convert_to_utf8(v)]
-          end
+        requests_to_send << {
+          'entries' => entries,
+          'log_name' => log_name,
+          'resource' => group_level_resource,
+          'labels' => group_level_common_labels
+        }
+      end
 
-          requests_to_send << Google::Logging::V2::WriteLogEntriesRequest.new(
-            entries: entries,
-            log_name: log_name,
-            resource: Google::Api::MonitoredResource.new(
-              type: group_level_resource.type,
-              labels: group_level_resource.labels.to_h
-            ),
-            labels: labels_utf8_pairs.to_h)
-        else
-          requests_to_send << \
-            Google::Apis::LoggingV2::WriteLogEntriesRequest.new(
-              log_name: log_name,
-              resource: group_level_resource,
-              labels: group_level_common_labels,
-              partial_success: @partial_success,
-              entries: entries)
+      if @shard_logs_by_tag
+        requests_to_send.each do |request|
+          @write_request.call(request)
         end
-      end
-      requests_to_send.each do |request|
-        client = api_client
-        if @use_grpc
-          write_request_via_grpc(client, request)
-        else
-          write_request_via_http(client, request)
+      else
+        # Combine all requests into one. The request level "log_name" will be
+        # ported to the entry level. The request level "resource" and "labels"
+        # are ignored as they should have been folded into the entry level
+        # "resource" and "labels" already anyway.
```

Comment:
Should mention somewhere that we're modifying the entries in-place...

Reply:
Changed.